plugins/fmhash/fmhash.c

```diff
@@ -0,0 +1,414 @@
+	if(func->nParams < 1) {
+		parser_errmsg("fmhash: hash64(string) / hash64(string, seed)"
+				" insufficient params. %d of file %s\n",
+			__LINE__, __FILE__);
+		iRet = RS_RET_ERR;
+		FINALIZE;
+	}
+	func->destructable_funcdata = 1;
+	CHKmalloc(hash_context = calloc(1, sizeof(hash_context_t)));
+	init_hash64_context(hash_context);
+	func->funcdata = (void*)&hash_context;
+
+finalize_it:
+	RETiRet;
+}
+
+static rsRetVal ATTR_NONNULL(1)
+init_fmHash64mod(struct cnffunc *const func)
+{
+	DEFiRet;
+	hash_context_t *hash_context = NULL;
+	if(func->nParams < 2) {
+		parser_errmsg("fmhash: hash64mod(string, mod)/hash64mod(string, mod, seed)"
+				" insufficient params. %d of file %s\n",
+			__LINE__, __FILE__);
+		iRet = RS_RET_ERR;
+		FINALIZE;
+	}
+	func->destructable_funcdata = 1;
+	CHKmalloc(hash_context = calloc(1, sizeof(hash_context_t)));
+	init_hash64_context(hash_context);
+	func->funcdata = (void*)&hash_context;
+finalize_it:
+	RETiRet;
+}
+
+static rsRetVal ATTR_NONNULL(1)
+init_fmHash32(struct cnffunc *const func)
+{
+	DEFiRet;
+	hash_context_t *hash_context = NULL;
+	if(func->nParams < 1) {
+		parser_errmsg("fmhash: hash64(string) / hash64(string, seed)"
+				" insufficient params. %d of file %s\n",
+			__LINE__, __FILE__);
+		iRet = RS_RET_ERR;
+		FINALIZE;
+	}
+	func->destructable_funcdata = 1;
+	CHKmalloc(hash_context = calloc(1, sizeof(hash_context_t)));
+	init_hash32_context(hash_context);
+	func->funcdata = (void*)&hash_context;
+
+finalize_it:
+	RETiRet;
+}
+
+static rsRetVal ATTR_NONNULL(1)
+init_fmHash32mod(struct cnffunc *const func)
+{
+	DEFiRet;
+	hash_context_t *hash_context = NULL;
+	if(func->nParams < 2) {
+		parser_errmsg("fmhash: hash64mod(string, mod)/hash64mod(string, mod, seed)"
+				" insufficient params. %d of file %s\n",
+			__LINE__, __FILE__);
+		iRet = RS_RET_ERR;
+		FINALIZE;
+	}
+	func->destructable_funcdata = 1;
+	CHKmalloc(hash_context = calloc(1, sizeof(hash_context_t)));
+	init_hash32_context(hash_context);
+	func->funcdata = (void*)&hash_context;
+finalize_it:
+	RETiRet;
+}
+
+static void ATTR_NONNULL(1)
+destruct_fmhash(struct cnffunc *const func)
+{
+	if(func->funcdata != NULL) {
+		if(((hash_context_t*)func->funcdata)->xhash != NULL) {
+			free((void*)((hash_context_t*)func->funcdata)->xhash);
+		}
+		free((void*) ((hash_context_t*)func->funcdata));
+	}
+}
+
+static struct scriptFunct functions[] = {
+	{"hash64", 1, 2, fmHashXX, init_fmHash64, destruct_fmhash},
+	{"hash64mod", 2, 3, fmHashXXmod, init_fmHash64mod, destruct_fmhash},
+	{"hash32", 1, 2, fmHashXX, init_fmHash32, destruct_fmhash},
+	{"hash32mod", 2, 3, fmHashXXmod, init_fmHash32mod, destruct_fmhash},
+	{NULL, 0, 0, NULL} //last element to check end of array
+};
+
+BEGINgetFunctArray
+CODESTARTgetFunctArray
+	dbgprintf("Hash: fmhhash\n");
+	*version = 1;
+	*functArray = functions;
+ENDgetFunctArray
+
+
+BEGINmodExit
+CODESTARTmodExit
+ENDmodExit
+
+#if !defined(__clang__)
+#pragma GCC diagnostic ignored "-Wstrict-prototypes"
```

Comment:
This looks strange to me, what is the error that comes up without this pragma? Maybe there is a bug in module-interface macros...

Reply:
http://build.rsyslog.com:8010/builders/rsyslog%20centos7%20rsyslog/builds/2936/steps/shell_2/logs/stdio

make[2]: Leaving directory `/var/lib/buildbot/slave/rsyslog_centos7_rsyslog/build/plugins/fmhttp'
Making all in plugins/fmhash
make[2]: Entering directory `/var/lib/buildbot/slave/rsyslog_centos7_rsyslog/build/plugins/fmhash'
  CC       fmhash_la-fmhash.lo
fmhash.c:396:1: error: function declaration isn't a prototype [-Werror=strict-prototypes]
 BEGINqueryEtryPt
 ^
fmhash.c:402:1: error: function declaration isn't a prototype [-Werror=strict-prototypes]
 BEGINmodInit()
 ^
fmhash.c:402:1: error: function declaration isn't a prototype [-Werror=strict-prototypes]
fmhash.c:402:1: error: function declaration isn't a prototype [-Werror=strict-prototypes]
fmhash.c:402:1: error: function declaration isn't a prototype [-Werror=strict-prototypes]
cc1: all warnings being treated as errors
make[2]: *** [fmhash_la-fmhash.lo] Error 1

Reply:
thx - that is strange. I will look into it. Please do not wonder if I "clone" your PR and submit it again - it then is just to have an example where this problem occurs. I'll delete mine later on.

Reply:
Sure, Thanks for your help.

Reply:
Please remove these two sets of pragmas. My test PR shows that the problem does no longer occur (checked centos 7 as in the sample). This is as expected. I guess the previous error was due to some other code that has been corrected in the mean time.